src/crypto.rs

```diff
@@ -25,40 +26,83 @@ const BLS_PUBLIC_KEY_BYTES_LEN: usize = 48;
 const BLS_SECRET_KEY_BYTES_LEN: usize = 32;
 const BLS_SIGNATURE_BYTES_LEN: usize = 96;
 
-#[derive(Debug, Error)]
+#[derive(Debug)]
 pub enum Error {
     #[cfg(feature = "serde")]
-    #[error("error deserializing hex-encoded input: {0}")]
-    Hex(#[from] HexError),
-    #[error("inputs required for aggregation but none were provided")]
+    Hex,
     EmptyAggregate,
-    #[error(
-        "invalid length of encoding: expected {expected} bytes but only provided {provided} bytes"
-    )]
-    EncodingError { provided: usize, expected: usize },
-    #[error("randomness failure: {0}")]
-    Randomness(#[from] rand::Error),
-    #[error("blst error: {0}")]
-    BLST(#[from] BLSTError),
-    #[error("invalid signature")]
+    EncodingError {
+        provided: usize,
+        expected: usize,
+    },
+    Randomness,
+    BLST,
     InvalidSignature,
+    KeyValidationFailed,
+}
+
+#[cfg(feature = "serde")]
+impl From<HexError> for Error {
+    fn from(_: HexError) -> Self {
+        Error::Hex
+    }
+}
+
+impl From<rand::Error> for Error {
+    fn from(_: rand::Error) -> Self {
+        Error::Randomness
+    }
+}
+
+impl From<BLSTError> for Error {
+    fn from(_: BLSTError) -> Self {
+        Error::BLST
+    }
```

Comment:
each of these error variants should grab the error they are wrapping

Reply:
Can you elaborate, I don't quite grasp what you mean here. Thanks